report/renderer/report-renderer.js

```diff
@@ -127,7 +127,7 @@ export class ReportRenderer {
       : 'Chromium';
     const channel = report.configSettings.channel;
     const benchmarkIndex = report.environment.benchmarkIndex.toFixed(0);
-    const axeVersion = report.environment.credits['axe-core'];
+    const axeVersion = report.environment.credits && report.environment.credits['axe-core'];
```

Comment:
```suggestion
    const axeVersion = report.environment.credits?.['axe-core'];
```

Reply:
Instead, shouldn't we fix this in `prepareReportResult`? I'd expect the renderer to break with older LHRs w/o `credits`.

Reply:
i mean... i'd love that, but don't we have an incompatibility in our tooling stack?

(perhaps it was browserify?)

even if we're 100% gtg here, i'd much rather do this `?.` upgrade in a bulk-change PR than introduce it with this one PR that i need for v9.

Reply:
before #13125 the renderer [also did a check](https://github.com/GoogleChrome/lighthouse/blob/031419ef2496a7349df7b3916edb2ed960a31d4e/report/renderer/report-renderer.js#L105) for `report.environment.credits` before accessing, so I'd personally be fine with keeping it local. `prepareReportResult` seems like it should be reserved for things we really need as opposed to more optional metadata.

We should probably [make the property optional](https://github.com/GoogleChrome/lighthouse/blob/665df67b3ec6fdc849ec646c7b388d3d2d4bc09f/types/lhr/lhr.d.ts#L63) to make sure it's always checked, though.

Reply:
> i mean... i'd love that, but don't we have an incompatibility in our tooling stack?
> 
> (perhaps it was browserify?)

yeah, but it would also still be fine because this is in the renderer, so it's browser only (and node for a unit test without any bundling)

> even if we're 100% gtg here, i'd much rather do this ?. upgrade in a bulk-change PR than introduce it with this one PR that i need for v9.

haha, a mass change sounds worse to me rather than starting to use it as needed :)

Reply:
> We should probably [make the property optional](https://github.com/GoogleChrome/lighthouse/blob/665df67b3ec6fdc849ec646c7b388d3d2d4bc09f/types/lhr/lhr.d.ts#L63) to make sure it's always checked, though.

this is basically [Report: add more robust handling of default property values stripped by LHR proto · #12868](https://github.com/GoogleChrome/lighthouse/issues/12868#issuecomment-893771191)

> it was browserify, but it would also still be fine because this is in the renderer, so it's browser only (and node for a unit test without any bundling)

k thanks for the confirmation.

Reply:
> > We should probably [make the property optional](https://github.com/GoogleChrome/lighthouse/blob/665df67b3ec6fdc849ec646c7b388d3d2d4bc09f/types/lhr/lhr.d.ts#L63) to make sure it's always checked, though.
> 
> this is basically [Report: add more robust handling of default property values stripped by LHR proto · #12868](https://github.com/GoogleChrome/lighthouse/issues/12868#issuecomment-893771191)

Yeah, fair point. I guess I was thinking differently about it because it was meant to be optionally handled from the start, so we should have just typed it like that instead of setting it as required but handling the optional fallback in the renderer. Without the type the next refactor was bound to remove the check.

We should probably write down guidance for ourselves for both robust handling of trips through proto space and easing compatibility with old LHRs.

Reply:
The separate consideration of some types to be optional on the LHR (when they really would always exist in the latest LH run) is confusing IMO. The "fixup" is simple enough (empty object for `credits`, done in the same place with fixup all other things about older LHRs before rendering), although for this property in particular that still leaves us needing to check the existence of the `axe-core` key (can we change the type to `Record<string, string|undefined>`?).